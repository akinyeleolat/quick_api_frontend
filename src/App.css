.navbar-container {
	min-height: 5rem;
	width: 100vw;
	background-color: #00bfff;
	display: flex;
	flex-direction: row;
	align-items: center;
	justify-content: flex-start;
	padding: 1rem 2.5vw;
}

.navbar-container .navbar-text {
	color: #fff;
	font-size: 1.2rem;
	font-weight: bold;
	margin: 0 4rem 0 0.5rem;
}
.navbar-container .navbar-icon {
	color: #fff;
	font-size: 1rem;
	margin: 0;
}

.navbar-container .navbar-title {
	color: white;
	font-size: 1.4rem;
	font-weight: 600;
}

.content-container {
	display: flex;
	flex-direction: row;
	align-items: center;
	flex-wrap: wrap;
	padding: 0rem 2.5vw;
	margin-top: 3rem;
}

.content-container .user-container {
	min-height: 200px;
	min-width: 170px;
	border-radius: 5px;
	margin: 1rem 1rem 1rem 0;
	padding: 0.5rem;
	border: 2px solid lightblue;
	display: flex;
	flex-direction: column;
}

.user-container .user-avatar {
	width: 4rem;
	height: 4rem;
	border-radius: 2rem;
	align-self: center;
	margin-top: 0.5rem;
}

.user-container .user-name {
	font-size: 1.1rem;
	font-weight: 600;
	margin-top: 1.5rem;
	text-align: center;
}

.user-container .buttons-container {
	display: flex;
	flex-direction: row;
	justify-content: space-around;
	margin-top: 1.5rem;
}

.add-box-container {
	min-height: 200px;
	min-width: 170px;
	border-radius: 5px;
	padding: 0.5rem;
	border: 2px solid lightblue;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	cursor: pointer;
}

.add-box-icon {
	font-size: 2.5rem;
	color: #00bfff;
}

.add-box-text {
	font-size: 1.1rem;
	font-weight: 600;
	margin-top: 1.2rem;
	text-align: center;
	color: #00bfff;
}

/* Create Users Page */

.page-container.form-page {
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
}

.form-container {
	max-width: 850px;
	margin-left: 3vw;
	border: 2px solid lightblue;
	padding: 2rem;
	margin-top: 4rem;
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
}

.form-container label {
	width: 50%;
	display: flex;
	flex-direction: column;
	font-size: 1.05rem;
	font-weight: 600;
}

.form-container label input {
	width: 90%;
	margin: 1rem 0;
	padding: 1rem;
	border-radius: 5px;
	border: 1px solid lightblue;
}

.submit-button {
	width: 45%;
	height: 3.5rem;
	border-radius: 5px;
	border: 2px solid #00bfff;
	font-size: 0.85rem;
	font-weight: 600;
	color: white;
	background-color: #00bfff;
	text-align: center;
	cursor: pointer;
	transition: all 0.45s ease;
	margin-top: 2rem;
}

.submit-button .fa-spinner{
	margin-left: 0.5rem;
	margin-right: 0.5rem;
	font-size: 1.1rem;
}

.submit-button:disabled {
	background-color: grey;
	color: white;
	border: none;
}

.submit-button:hover {
	box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.25);
	transform: translateY(-0.25rem);
}

.fa-spin {
	font-size: 2rem;
}

/*  */

.table-container {
	overflow-x: scroll;
}

.table-heading-container {
	display: flex;
	flex-direction: row;
	justify-content: space-around;
	align-items: center;
	margin-bottom: 1rem;
	width: 100vw;
}

.table-heading {
	font-size: 1.5rem;
	font-weight: 600;
	color: #00bfff;
}

.table-container::-webkit-scrollbar {
	height: 8px;
	background-color: lightblue;
	padding: 3px;
}

.table-container::-webkit-scrollbar-thumb {
	background-color: grey;
	border-radius: 5px;
}

table {
	border-collapse: collapse;
	border-spacing: 0;
	width: 100%;
	border: 1px solid #ddd;
}

th,
td {
	text-align: left;
	padding: 8px 7px;
	font-size: 0.9rem;
	height: 3.5rem;
	width: 10.5vw;
	min-width: 12rem;
	max-width: 15rem;
}

tr:nth-child(even) {
	background-color: #f2f2f2;
}

tr td:last-of-type,
tr th:last-of-type {
	min-width: 16rem; /* This is the minimum width of the last column to give space for floating actions */
}

tr .floating-actions {
	position: absolute;
	right: 0;
	width: 5rem;
	right: calc(3vw - 10px);
	background-color: white;
	height: 3.4rem;
	margin-top: 0.05rem;
	padding: 8px 10px;
	display: flex;
	align-items: center;
	justify-content: space-around;
}

.floating-actions i {
	font-size: 1.1rem;
	color: #00bfff;
	cursor: pointer;
}

.floating-actions i.red {
	color: red;
}

.table-button {
	padding: 0.5rem 1.5rem;
	margin: auto;
	margin-top: 2rem;
	font-size: 1rem;
	background-color: #00bfff;
	color: white;
	cursor: pointer;
	border-radius: 5px;
	font-weight: 500;
}