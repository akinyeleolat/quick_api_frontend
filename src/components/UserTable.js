import {Link} from 'react-router-dom'
import React, {useEffect, useState} from 'react'
import {useNavigate} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux"
import {getAllUser, removeUser} from '../store/actions';


export default function UserTable({data, ...props}) {
    const navigate = useNavigate()
    const userList = useSelector(state => state.app?.users)
    const dispatch = useDispatch()
    const [loading, setLoading] = useState(true)


    const showTableHeading = () => {
        if (loading) {
            return <i className='fa fa-spinner fa-spin'></i>
        } else if (!loading && userList?.length === 0) {
            return "No Users Found"
        } else {
            return userList?.length + " Users"
        }
    }

    const getUsersFromAPI = async () => {
        setLoading(true)
        await dispatch(getAllUser())
        setLoading(false)
    }



    const handleDelete = (id) => {
        let shouldDelete = window.confirm('Are you sure you want to delete this user?')
        if (shouldDelete) {
            dispatch(removeUser(id))
                .then(data => getUsersFromAPI())
                .catch(err => alert(err.message || err || "Something went wrong deleting the user"))
        }
    }

    const handleEdit = (user) => {
        navigate('/edit-user', {state: {user}})
    }


    useEffect(() => {
        getUsersFromAPI()
    }, [])


    return (
        <>
            <div className="table-heading-container">
                <div className='table-heading'>User List</div>
                <div className=''>
                    {showTableHeading()}
                </div>
            </div>
            <div className='table-container'>
                <table data-testid="user-table">
                    <thead>
                    <tr>
                        <th>First Name</th>
                        <th>Last Name</th>
                        <th>Address 1</th>
                        <th>Address 2</th>
                        <th>Town</th>
                        <th>Region</th>
                        <th>Country</th>
                        <th>Postcode</th>
                        <th>Contact Number</th>
                    </tr>
                    </thead>
                    <tbody>
                    {userList.map((user, index) => (
                        <tr key={index.toString()}>
                            <td>{user.firstName}</td>
                            <td>{user.lastName}</td>
                            <td>{user.address_1}</td>
                            <td>{user.address_2}</td>
                            <td>{user.town}</td>
                            <td>{user.region}</td>
                            <td>{user.country}</td>
                            <td>{user.postCode}</td>
                            <td>{user.contactNumber}</td>
                            <div className='floating-actions'>
                                <i className="fa fa-pencil-square-o"
                                    aria-hidden="true"
                                    onClick={() => handleEdit(user)}></i>
                                <i className="fa fa-trash red" aria-hidden="true" onClick={() => handleDelete(user)}></i>
                            </div>
                        </tr>
                    ))}
                    {userList.length === 0 ? <tr>
                        <td colSpan={8}>No Users Found</td>
                    </tr> : null}
                    </tbody>
                </table>
            </div>
            <Link to='/create-user' className='table-button'>
                <div>Create New User</div>
            </Link>
        </>
    )
}

