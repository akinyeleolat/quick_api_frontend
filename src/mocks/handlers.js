import {rest} from "msw";

export const handlers = [
  rest.get("https://us-central1-haud-test-api.cloudfunctions.net/app/api/getuser", (req, res, ctx) => {
    return res(
      ctx.json({
        users: [
          {
            firstName: "Firstname",
            lastName: "Lastname",
            address1: "Address1",
            address2: "Address2",
            town: "Town",
            region: "Region",
            postcode: "Postcode",
            contactNumber: "ContactNumber",
          },
        ],
        info: {
          seed: "1eefa7d651be03bf",
          results: 1,
          page: 1,
          version: "1.3",
        },
      })
    );
  }),
  rest.get("https://us-central1-haud-test-api.cloudfunctions.net/app/api/getuser/test_id", (req, res, ctx) => {
    return res(
      ctx.json({
        user:{
            firstName: "Firstname",
            lastName: "Lastname",
            address1: "Address1",
            address2: "Address2",
            town: "Town",
            region: "Region",
            postcode: "Postcode",
            contactNumber: "ContactNumber",
          },
        info: {
          seed: "1eefa7d651be03bf",
          results: 1,
          page: 1,
          version: "1.3",
        },
      })
    );
  }),
];
