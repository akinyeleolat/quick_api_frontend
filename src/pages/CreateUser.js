import axios from 'axios';
import React, {useState} from 'react'
import {Link, useLocation} from "react-router-dom";
import {useNavigate} from "react-router-dom";
import {useDispatch } from "react-redux";
import { createUser } from '../store/actions';

export default function CreateUsers({navigation, ...props}) {
    const navigate = useNavigate()
    const dispatch = useDispatch()
    const [loading, setLoading] = useState(false)
    const [inputValues, setInputValues] = useState({
        firstName: '',
        lastName: '',
        address_1: '',
        address_2: '',
        town: '',
        region: '',
        country: '',
        postCode: '',
        contactNumber: '',
    })


    const handleChange = (event) => {
        const {name, value} = event.target
        setInputValues({...inputValues, [name]: value})
    }


    const handleSubmit = async (event) => {
        setLoading(true)
        event.preventDefault()
       await dispatch(createUser(inputValues))
            .then(data => {
                console.log(data)
                setLoading(false)
                alert("User created successfully")
                navigate('/')
            }).catch(err => alert(err.message || err || "Something went wrong creating the user"))
        setLoading(false)
    }


    return (
        <div className="page-container form-page">
            <div className='navbar-container'>
                <Link to="/">
                    <div className="row">
                        <i className='fa fa-arrow-left navbar-icon'> </i>
                        <div className='navbar-text'>Back</div>
                    </div>
                </Link>
                <div className="navbar-title" data-testid="navbar-title">Create User</div>
            </div>
            <form className='form-container' onSubmit={handleSubmit}>
                <label>
                    First Name: <input
                        type="text"
                        name="firstName"
                        required
                        value={inputValues.firstName}
                        onChange={handleChange} />
                </label>
                <label>
                    Last Name: <input
                        type="text"
                        name="lastName"
                        required
                        value={inputValues.lastName}
                        onChange={handleChange} />
                </label>
                <label>
                    Address One: <input
                        type="address"
                        name="address_1"
                        value={inputValues.address_1}
                        onChange={handleChange} />
                </label>
                <label>
                    Address Two: <input
                        type="address"
                        name="address_2"
                        value={inputValues.address_2}
                        onChange={handleChange} />
                </label>
                <label>
                    Town: <input
                        type="text"
                        name="town"
                        value={inputValues.town}
                        onChange={handleChange} />
                </label>
                <label>
                    Region: <input
                        type="text"
                        name="region"
                        value={inputValues.region}
                        onChange={handleChange} />
                </label>
                <label>
                    Country: <input
                        type="text"
                        name="country"
                        value={inputValues.country}
                        onChange={handleChange} />
                </label>
                <label>
                    PostCode: <input
                        type="text"
                        name="postCode"
                        value={inputValues.postCode}
                        onChange={handleChange} />
                </label>
                <label>
                    Contact Number: <input
                        type="text"
                        name="contactNumber"
                        value={inputValues.contactNumber}
                        onChange={handleChange} />
                </label>
                <button className="submit-button red">
                    {loading ? "Creating" : "Create User"}
                    {loading ? <i className="fas fa-spinner fa-spin"></i> : null}
                </button>
                <Link to="/">
                    <div className="row">
                        <i className='fa fa-arrow-left'></i>
                        <div className='navbar-text'>Cancel</div>
                    </div>
                </Link>
            </form>
        </div>
    )
}
