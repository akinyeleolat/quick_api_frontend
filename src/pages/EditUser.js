import React, {useState} from 'react'
import {Link, useLocation} from "react-router-dom";
import {useNavigate} from "react-router-dom";
import {useDispatch } from "react-redux";
import { updateUser, getUser } from '../store/actions';


export default function EditUsers(props) {
    const params = useLocation();
    const dispatch = useDispatch()
    const [user, setUser] = useState(params.state?.user);
    const navigate = useNavigate()
    const [loading, setLoading] = useState(false)

    const [inputValues, setInputValues] = useState({
        firstName: user.firstName || '',
        lastName: user.lastName || '',
        address_1: user.address_1 || '',
        address_2: user.address_2 || '',
        town: user.town || '',
        region: user.region || '',
        country: user.country || '',
        postCode: user.postCode || '',
        contactNumber: user.contactNumber || '',
        id: user.id || ''
    })


    const getCurrentUser = async () => {
       await dispatch(getUser(user.id))
            .then(data => {
                console.log(data?.data?.data, "returned data")
                setInputValues({
                    firstName: data?.data?.data?.firstName || '',
                    lastName: data?.data?.data?.lastName || '',
                    address_1: data?.data?.data?.address_1 || '',
                    address_2: data?.data?.data?.address_2 || '',
                    town: data?.data?.data?.town || '',
                    region: data?.data?.data?.region || '',
                    country: data?.data?.data?.country || '',
                    postCode: data?.data?.data?.postCode || '',
                    contactNumber: data?.data?.data?.contactNumber || '',
                    id: data?.data?.data?.id || ''
                })
                setUser({
                    firstName: data?.data?.data?.firstName || '',
                    lastName: data?.data?.data?.lastName || '',
                    address_1: data?.data?.data?.address_1 || '',
                    address_2: data?.data?.data?.address_2 || '',
                    town: data?.data?.data?.town || '',
                    region: data?.data?.data?.region || '',
                    country: data?.data?.data?.country || '',
                    postCode: data?.data?.data?.postCode || '',
                    contactNumber: data?.data?.data?.contactNumber || '',
                    id: data?.data?.data?.id || ''
                })
            }).catch(err => {
                navigate('/')
                alert(err.message || err || "Something went wrong geting the updated user")
            })
    }

    const handleChange = (event) => {
        const {name, value} = event.target
        setInputValues({...inputValues, [name]: value})
    }


    const checkDisabled = () => {
        const isEqual = Object.keys(inputValues).every(key => inputValues[key] === user[key])
        if (isEqual) {
            return true
        } else {
            return false
        }
    }


    const handleSubmit = async (event) => {
        setLoading(true)
        event.preventDefault()
        dispatch(updateUser(inputValues,user.id))
            .then(data => {
                setLoading(false)
                getCurrentUser()
                setLoading(false)
                alert("User updated successfully")
                navigate('/')
            }).catch(err => alert(err.message || err || "Something went wrong updating the user"))
        setLoading(false)
    }

    return (
        <div className="page-container form-page">
            <div className='navbar-container'>
                <Link to="/">
                    <div className="row">
                        <i className='fa fa-arrow-left navbar-icon'> </i>
                        <div className='navbar-text'>Back</div>
                    </div>
                </Link>
                <div className="navbar-title" data-testid="navbar-title">Edit User</div>
            </div>
            <form className='form-container' onSubmit={handleSubmit}>
                <label>
                    First Name: <input
                        type="text"
                        name="firstName"
                        required
                        data-testid="firstname-input"
                        value={inputValues.firstName}
                        onChange={handleChange} />
                </label>
                <label>
                    Last Name: <input
                        type="text"
                        name="lastName"
                        required
                        data-testid="lastname-input"
                        value={inputValues.lastName}
                        onChange={handleChange} />
                </label>
                <label>
                    Address One: <input
                        type="address"
                        name="address_1"
                        value={inputValues.address_1}
                        data-testid="address1-input"
                        onChange={handleChange} />
                </label>
                <label>
                    Address Two: <input
                        type="address"
                        name="address_2"
                        value={inputValues.address_2}
                        data-testid="address2-input"
                        onChange={handleChange} />
                </label>
                <label>
                    Town: <input
                        type="text"
                        name="town"
                        value={inputValues.town}
                        data-testid="town-input"
                        onChange={handleChange} />
                </label>
                <label>
                    Region: <input
                        type="text"
                        name="region"
                        data-testid="region-input"
                        value={inputValues.region}
                        onChange={handleChange} />
                </label>
                <label>
                    Country: <input
                        type="text"
                        name="country"
                        data-testid="country-input"
                        value={inputValues.country}
                        onChange={handleChange} />
                </label>
                <label>
                    PostCode: <input
                        type="text"
                        name="postCode"
                        data-testid="postcode-input"
                        value={inputValues.postCode}
                        onChange={handleChange} />
                </label>
                <label>
                    Contact Number: <input
                        type="text"
                        name="contactNumber"
                        data-testid="contact-number-input"
                        value={inputValues.contactNumber}
                        onChange={handleChange} />
                </label>
                <button className="submit-button red loading" disabled={checkDisabled()}>
                    {loading ? "Updating" : "Update User"}
                    {loading ? <i className="fas fa-spinner fa-spin"></i> : null}
                </button>
            </form>
        </div>
    )
}
