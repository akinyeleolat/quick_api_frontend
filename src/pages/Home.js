import React from 'react'
import UserTable from '../components/UserTable';


export default function Home(props) {
    
    return (
        <div className="page-container">
            <div className='navbar-container'>
                <div className="navbar-title" data-testid="navbar-title">Home</div>
            </div>
            <div className='content-container'>
                {/* Render All Users */}
                <UserTable />
            </div>
        </div>
    )
}
