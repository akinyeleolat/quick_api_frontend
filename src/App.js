import React from 'react';
import {
  BrowserRouter,
  Routes,
  Route
} from "react-router-dom";
// import your route components too

// import Screens
import Home from './pages/Home';
import EditUser from './pages/EditUser';
import CreateUser from './pages/CreateUser';

import "./App.css";

export default function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/edit-user" element={<EditUser />} />
        <Route path="create-user" element={<CreateUser />} />
      </Routes>
    </BrowserRouter>
  )
}