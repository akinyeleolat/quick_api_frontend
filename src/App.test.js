import {
  fireEvent,
  render,
  screen,
  waitFor
} from "@testing-library/react";
import {act} from "react-dom/test-utils";
import Home from "./pages/Home";
import App from "./App";
import CreateUser from "./pages/CreateUser";
import EditUser from "./pages/EditUser";
import {server} from "./mocks/server";
import {MemoryRouter} from "react-router-dom";
import {Provider} from 'react-redux'
import thunk from "redux-thunk";
import configureMockStore from 'redux-mock-store';
import UserTable from "./components/UserTable";



beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());


describe("AppTest", () => {
  // mock store for testing
  const middlewares = [thunk]
  const initialState = {app: {users: []}};
  const mockStore = configureMockStore(middlewares)
  let store = mockStore(initialState)

  it("Renders react app without crashing", async () => {
    act(() => {
      render(
        <Provider store={store}>
          <App />
        </Provider>
      );
    });
  });


  it("Renders the home page correctly", async () => {
    let homeHeader;
    act(() => {
      const {getByTestId} = render(
        <Provider store={store}>
          <MemoryRouter initialEntries={["/"]}>
            <Home>
              <UserTable/>
              </Home>
          </MemoryRouter>
        </Provider>
      );
      homeHeader = getByTestId("navbar-title");
    });
    await expect(homeHeader.textContent).toBe(`Home`);
  });


  it("Renders the home page correctly with the usertable", async () => {
    let userTable;
   await act(async () => {
      const {getByTestId, asFragment} = render(
        <Provider store={store}>
          <MemoryRouter initialEntries={["/"]}>
            <Home>
              <UserTable/>
              </Home>
          </MemoryRouter>
        </Provider>
      );
      userTable = await waitFor(()=>{
        return getByTestId("user-table");
      });
    expect(userTable.children).toHaveLength(2);
    expect(asFragment()).toMatchSnapshot();
    });
    
  });
 


  it("Renders the create user page correctly", async () => {
    let createUserHeader;
    act(() => {
      const {getByTestId} = render(
        <Provider store={store}>
          <MemoryRouter initialEntries={["/create-user"]}>
            <CreateUser />
          </MemoryRouter>
        </Provider>
      );
      createUserHeader = getByTestId("navbar-title");
    });
    await expect(createUserHeader.textContent).toBe(`Create User`);
  });

});