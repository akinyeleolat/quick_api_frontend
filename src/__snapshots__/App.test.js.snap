// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTest Renders the home page correctly with the usertable 1`] = `
<DocumentFragment>
  <div
    class="page-container"
  >
    <div
      class="navbar-container"
    >
      <div
        class="navbar-title"
        data-testid="navbar-title"
      >
        Home
      </div>
    </div>
    <div
      class="content-container"
    >
      <div
        class="table-heading-container"
      >
        <div
          class="table-heading"
        >
          User List
        </div>
        <div
          class=""
        >
          <i
            class="fa fa-spinner fa-spin"
          />
        </div>
      </div>
      <div
        class="table-container"
      >
        <table
          data-testid="user-table"
        >
          <thead>
            <tr>
              <th>
                First Name
              </th>
              <th>
                Last Name
              </th>
              <th>
                Address 1
              </th>
              <th>
                Address 2
              </th>
              <th>
                Town
              </th>
              <th>
                Region
              </th>
              <th>
                Country
              </th>
              <th>
                Postcode
              </th>
              <th>
                Contact Number
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                colspan="8"
              >
                No Users Found
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <a
        class="table-button"
        href="/create-user"
      >
        <div>
          Create New User
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
